fix remember me login

ChristophWurst opened the pull request:
fixes https://github.com/nextcloud/server/issues/1067

I somehow refactored away the remember-me funcionality while working on the auth code :see_no_evil: 

TODO:
- [x] ~~update session token in the DB~~ duplicate token, otherwise the validation fails and the user is logged out
- [x] only set remember cookies when the 2FA challenge was solved successfully
- [x] only set remember cookie when the checkbox was checked on the login page
- [x] flag remember-me session tokens and don't delete them after 24h, but rather after the max remember-me lifetime elapsed, relative to the last usage time
- [x] tests

cc @LukasReschke

ChristophWurst committed:
bring back remember-me
* try to reuse the old session token for remember me login
* decrypt/encrypt token password and set the session id accordingly
* create remember-me cookies only if checkbox is checked and 2fa solved
* adjust db token cleanup to store remembered tokens longer
* adjust unit tests

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

coderkun commented:
`if (true) {` ?

ChristophWurst commented:
> if (true) { ?

This is where I'll have to check whether the checkbox was checked or not :wink:

MeCias commented:
Hi,

Maybe you can consider [this request](https://github.com/nextcloud/server/issues/1358) as well.

Thanks

ChristophWurst commented:
The basic remember-me-login works again, but the session token validation logic logs the user out because it cannot find a valid session token in the DB for the current session id on the next request.

Solution 1: Re-use old session token
We could update an existing session token with the new session id. However, we'd have to tweak the TTL of 24h, because currently we delete old tokens in a background job. Additionally, concurrency will be tricky, e.g. handling two or more concurrent requests that should be logged in via remember me – which one gets the new session id?

Solution 2: Create a new session token
Maybe easier to implement, but we wouldn't have a password for the new token. That means we can no longer rely on the *kill existing connections by changing the password* as sessions without a password obviously don't use that logic. Concurrency might also be a problem here, as concurrent requests might result into multiple new session tokens. Additionally, the list of active sessions is expanded if the session is lost within the 24h time span, so this might be confusing to see multiple dead connections in the personal settings.

I've thought about this a log, but have not come to a clean, robust solution. @LukasReschke @MorrisJobke @rullzer @icewind1991 does anybody have an idea how we could fix that in the scope of 10.0.2?

ChristophWurst commented:
Updated solution 1: Construct a new session token with the data from the old one.

That means we'd have to store the session token ID or similar next to the remember me cookie (either additional cookie or combine somehow). When we try to log in by cookie, we first run the old logic (lookup login token in DB). Then we load the old session token from the database, based on the value of the remember me cookie. That old session token will be cloned (create a new one with same attributes, but with the new session ID of course). Future requests should be successfully authenticated again.

We can then safely delete the old token.

Additional problem: we have to tweak the lifetime of session tokens in the DB because they are currently deleted after 24h of inactivity.

@LukasReschke @rullzer @nickvergessen @MorrisJobke objections? I'd try to implement this if you agree that this is the way to go.

ChristophWurst commented:
Fixed and added some tests, squashed my commits and rebased to resolve some conflicts.

@LukasReschke @MorrisJobke @rullzer this is ready for testing.

I faked a lost session by deleting the session cookie and php's session files.

ChristophWurst commented:
some tests are still failing

rullzer commented on lib/private/User/Session.php:
```diff
@@ -710,6 +723,15 @@ public function loginWithCookie($uid, $currentToken) {
+	 * @param IUser $user
+	 */
+	public function createRememberMeToken(IUser $user) {
+		$token = OC::$server->getSecureRandom()->generate(32);
```
Why not inject this?

ChristophWurst replied:
fixed

rullzer commented on lib/private/User/Session.php:
```diff
@@ -736,10 +758,19 @@ public function logout() {
+			$webRoot = '/';
+		}
+
+		$expires = $this->timeFacory->getTime() + OC::$server->getConfig()->getSystemValue('remember_login_cookie_lifetime', 60 * 60 * 24 * 15);
```
Config is injected right?

ChristophWurst replied:
yes, it is. fixed it and also replaced all ``time()`` calls with the injected time factory

ChristophWurst committed:
inject ISecureRandom into user session and use injected config too

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
rebased to resolve session test conflicts.

@rullzer @LukasReschke @nickvergessen review please :-)

nickvergessen commented on core/Controller/TwoFactorChallengeController.php:
```diff
@@ -32,14 +33,14 @@
 use OCP\ISession;
 use OCP\IURLGenerator;
-use OCP\IUserSession;
+use OCP\IUserSession as UserSession;
```
Any reason for this?

ChristophWurst replied:
no. I forgot to revert this. removed it.

ChristophWurst commented:
```
  - git@github.com:fruux/sabre-dav.git                                                                                                        
    Cloning into '/drone/src/github.com/nextcloud/server/build/integration/vendor/sabre/dav'...                                               
    Host key verification failed.                                                                                                             
    fatal: Could not read from remote repository.                                                                                             
                                                                                                                                              
    Please make sure you have the correct access rights                                                                                       
    Please make sure you have the correct access rights                                                                                       
```
@MorrisJobke any idea what's failing there?

MorrisJobke commented:
> @MorrisJobke any idea what's failing there?

Because you are only allowed to clone authenticated via git@github.com - You should use git clone https://github.com/... for anonymous cloning. ;)

MorrisJobke commented:
I rebased - maybe this helps.

rullzer commented:
Looks good to me. Also works again over here. We might miss a corner case somewhere but LGTM!

LukasReschke commented:
Will take a look at this as well.

LukasReschke commented on lib/private/Authentication/Token/DefaultTokenProvider.php:
```diff
@@ -152,6 +154,27 @@ public function getToken($tokenId) {
+	 * @param string $oldSessionId
+	 * @param string $sessionId
+	 */
+	public function renewSessionToken($oldSessionId, $sessionId) {
```
Function needs tests. Will write myself later.

LukasReschke commented on lib/private/Authentication/TwoFactorAuth/Manager.php:
```diff
@@ -37,6 +37,7 @@ class Manager {
 	const SESSION_UID_KEY = 'two_factor_auth_uid';
 	const BACKUP_CODES_APP_ID = 'twofactor_backupcodes';
 	const BACKUP_CODES_PROVIDER_ID = 'backup_codes';
+	const REMEBER_LOGIN = 'two_factor_remember_login';
```
Typo in the constant name

LukasReschke commented on lib/private/User/Session.php:
```diff
@@ -710,6 +730,15 @@ public function loginWithCookie($uid, $currentToken) {
 	/**
+	 * @param IUser $user
+	 */
+	public function createRememberMeToken(IUser $user) {
```
Needs tests

LukasReschke replied:
Added.

LukasReschke reviewed (changes requested):
Looks generally good. Will add some missing tests and also test locally.

LukasReschke commented on lib/private/Authentication/Token/IProvider.php:
```diff
@@ -51,6 +53,12 @@ public function generateToken($token, $uid, $loginName, $password, $name, $type
 	public function getToken($tokenId) ;
 
 	/**
+	 * @param string $oldSessionId
```
Add at least a single line comment what the function is supposed to do? :)

ChristophWurst replied:
done

LukasReschke commented on lib/private/Authentication/TwoFactorAuth/Manager.php:
```diff
@@ -51,6 +52,7 @@ class Manager {
 	 * @param AppManager $appManager
 	 * @param ISession $session
 	 * @param IConfig $config
+	 * @param Session $userSession
```
We're passing a ISession and not a Session

LukasReschke replied:
Actually it seems like that is even not ever used :)

LukasReschke committed:
Fix invalid PHPDocs

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke committed:
Fix typ in constant name

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke committed:
Add missing tests and fix PHPDoc

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

ChristophWurst committed:
document what the method does

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
> LukasReschke requested changes

@LukasReschke anything left?

LukasReschke commented on .htaccess:
```diff
@@ -72,3 +72,7 @@ Options -Indexes
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 /core/templates/403.php
+ErrorDocument 404 /core/templates/404.php
```
> @LukasReschke anything left?

Yes.

ChristophWurst replied:
fixed

LukasReschke commented:
LGTM once the last issue is fixed :)

chaos-prevails commented:
Hello,

I'm using NC11 just downloaded yesterday and I want to disable saving the login session cookie. In fact, I want that users get automatically logged out after 10 minutes of inactivity.

This works fine:
'remember_login_cookie_lifetime' => 0,
'session_lifetime' => 600,
'session_keepalive' => false,

But the "stay logged in" checkbox on the login screen is still visible (although disabled by 'remember_login_cookie_lifetime' => 0 AFAIK).

The addon "disable remember login" (from OC) is not available any more in the app store of NC11, so I assume (wrongly?) I shouldn't use it any more with NC11 (was available with NC10).

Is there a way to hide the checkbox on the login screen? Following this thread I assumed I have to set a config.php parameter but I don't see which one.

thanks alot,
Tim

My1 commented:
@chaos-prevails just enable the external storage app but in the admin interface do not give the users the permission to add external storages, this should take care of the remember login checkbox.

chaos-prevails commented:
Thanks, that did the trick. However, even when disabling the ability to add external storage, the "external storage" link at the left side menu shows up for all users.

Probably there is a more direct way to hide the checkbox? I might rather try the old app on NC11.

My1 commented:
well, if I would know how the external storage kicks out the checkbox we could solve both our problems.
I could get external storage and remember me, and you could for example create a dummy plugin which does nothing except for killing off the checkbox.

otherwise you may play with the login page files and remove anything regarding the checkbox.

chaos-prevails commented:
I'm not sure whether the checkbox is disappearing without reason when enabling the external storage app. probably it's not possible any more due to the login necessities to the external storage sites? (I have no idea, sorry). 

I will try out the old app to check whether it works on NC11

MorrisJobke commented:
> well, if I would know how the external storage kicks out the checkbox we could solve both our problems.

This is how it kicks out the checkbox: https://github.com/nextcloud/server/blob/bc26f78eb1d87e7e9dfa62cf97e8a0cb81f79fa2/apps/files_external/appinfo/info.xml#L15 😉

My1 commented:
oh my god, why did I overlook that. I did check the metafiles iirc. thanks.

chaos-prevails commented:
I have to revert my answer from yesterday, actually none of the two things I try to achieve work out:

1) expire NC session when closing browser (independent from session-timeout)
does work in Chromium, does NOT work in Firefox

2) expire NC session after session timeout
does work for 10 seconds (testing), but doesn't work any more for e.g. 1 minute. I have switched to another tab to make sure I don't trigger any ajax calls, whatever. Still, I'm not disconnected after more than a 1 minute.

my config settings:
'remember_login_cookie_lifetime' => 0,
'session_lifetime' => 60,
'session_keepalive' => false,

beside client/browser specific issues (private browsing, etc), is there something else I have to consider so that it works out of the box browser-independently?

thanks
Tim

btw: I deleted all old cookies before testing and hide the remember login checkbox with the help of @MorrisJobke yesterday.
